DOC: Chart Using PyGal added (#762)
PyFPDF/fpdf2: docs/Maths.md

![](plotly_svg.png)

## Using Pygal ##

> [Pygal](https://www.pygal.org/en/stable/) is a graph plotting library using python. You can install Pygal using ` $ pip install pygal` command. 

Reviewer: ```suggestion
[Pygal](https://www.pygal.org/en/stable/) is a graph plotting library using Python. You can install Pygal using `pip install pygal` command. 
```

> `fpdf2` is able to embed the graph and charts that are generated using `Pygal` library. The following ways explain how to embed `Pygal` charts into `fpdf2` library.

Reviewer: Could you please add a sentence there explaining why we cannot embed graphs as SVG directly,
as we discussed in https://github.com/PyFPDF/fpdf2/issues/652#issuecomment-1501601977? 😊

Author: OK @Lucas-C . I will add and commit this ASAP.

> ### Using cairosvg (*A faster and efficient implementation*) ### 

Reviewer 2: Your examples convert the SVG output of pygal to a raster file (PNG) before passing it to fpdf2, even with two different methods and three seperate third-party libraries.
Why?
The resulting PDF file will be much larger and the diagram can't be scaled losslessly anymore. For the user of the produced file this is very inefficient and causes a severe loss of quality.

It would make a lot more sense to just show how to feed the SVG from PyGal directly to fpdf2. After all, this capability is one of the key features that distinguishes fpdf2 from other PDF libraries.

A comparison of how otherwise unrelated libraries can convert SVG data to raster images does not really seem to be a relevant topic for the fpdf2 documentation.

Reviewer: @gmischler this was already discussed in https://github.com/PyFPDF/fpdf2/issues/652#issuecomment-1501601977

Have you read the discussion in this issue? 😊

Reviewer 2: Ah I see. That's a rather annoying way to create SVGs (even referencing online fonts). I'm sure it makes it much simpler for them, though... :wink:

In that case, maybe the example should actually mention that the SVG files generated by PyGal can't be used directly because of the special features they use. The current phrasing makes the reader think that the files could be imported directly, but then without explanation does it in a different way.

Thinking about it some more, the general approach shown is not really specific to PyGal. Maybe this information should rather be titled "Importing Graphics from Incompatible SVG Data", with PyGal used as an example? It could even go to the "SVG.md" page, because it doesn't only apply to graphs.

Reviewer: > In that case, maybe the example should actually mention that the SVG files generated by PyGal can't be used directly because of the special features they use. The current phrasing makes the reader think that the files could be imported directly, but then without explanation does it in a different way.

That's exactly what I suggested: https://github.com/PyFPDF/fpdf2/pull/762#discussion_r1162460377

Author: Hello @gmischler , thank you so much for reviewing this PR and sharing valuable insights. 
I tried to do some research on my own and was not clear about the reason behind not being able to import incompatible SVG data into `fpdf2` until @Lucas-C shared it with me. 
I will try to reflect on the suggestions in my next commit. 
And yes,
> After all, this capability is one of the key features that distinguishes fpdf2 from other PDF libraries

I am aligned with the same thought. Being a beginner in open-source contribution, I believe I am learning every single day and I wish to become an active contributor like you people. 

@Lucas-C I will try to follow the advice as much as I can.

A faster and expected approach of embedding a `Pygal` svg graph into a PDF file is to use the `cairosvg` library to convert the `svg` string generated by `pygal` into byte string using `BytesIO` library so that we can keep these data in an in-memory buffer. 
As the `fpdf` library can understand byte string of a `svg` content, it can easily embed a graph inside a `pdf`. 

```python
import pygal
from fpdf import FPDF
from io import BytesIO
import cairosvg

# Create a Pygal bar chart
bar_chart = pygal.Bar()
bar_chart.title = 'Browser usage evolution (in %)'
bar_chart.x_labels = map(str, range(2002, 2013))
bar_chart.add('Firefox', [None, None, 0, 16.6,   25,   31, 36.4, 45.5, 46.3, 42.8, 37.1])
bar_chart.add('Chrome',  [None, None, None, None, None, None,    0,  3.9, 10.8, 23.8, 35.3])
bar_chart.add('IE',      [85.8, 84.6, 84.7, 74.5,   66, 58.6, 54.7, 44.8, 36.2, 26.6, 20.1])

> bar_chart.add('Others',  [14.2, 15.4, 15.3,  8.9,    9, 10.4,  8.9,  5.8,  6.7,  6.8,  7.5])

Reviewer: You used a very nice, clearly structured dataset as example! Well done 👍

Author: Thank You.

# Use CairoSVG to convert PNG from SVG of barchart

> svg_img_byte = BytesIO()

Reviewer: ```suggestion
svg_img_bytesio = BytesIO()
```
This is just a suggestion: do you think this naming would be clearer?

Author: Right. Thanks for the review. I will edit and commit this change soon.

cairosvg.svg2png(bar_chart.render(), write_to=svg_img_byte)

# Set the position and size of the image in the PDF
x = 50
y = 50
w = 100
h = 70

# Make the PDF
pdf = FPDF()
pdf.add_page()
pdf.image(svg_img_byte, x=x, y=y, w=w, h=h)
pdf.output('bar_chart.pdf')
```
The above code generates a pdf with title `bar_chart.pdf` file with following graph -
![](pygal_chart_cairo.png)

**!! Troubleshoot: !!**

You may encounter `GTK` (Gnome Toolkit) errors while executing the above example in windows. Error could be like following -
```
OSError: no library called "cairo-2" was found
no library called "cairo" was found
no library called "libcairo-2" was found
cannot load library 'libcairo.so.2': error 0x7e
cannot load library 'libcairo.2.dylib': error 0x7e
cannot load library 'libcairo-2.dll': error 0x7e
```

> In this case install install `GTK` from [GTK-for-Windows-Runtime-Environment-Installer](https://github.com/tschoonj/GTK-for-Windows-Runtime-Environment-Installer/releases). Restart your editor. And you are all done.

Reviewer: 👍 Excellent, nice addition!

### Using svglib and reportlab (*A slower and purely pythonic implementation*) ### 
```python
import pygal
from reportlab.graphics import renderPM
from svglib.svglib import svg2rlg
from fpdf import FPDF

# Create a Pygal bar chart
bar_chart = pygal.Bar()
bar_chart.title = 'Sales by Year'
bar_chart.x_labels = ['2016', '2017', '2018', '2019', '2020']
bar_chart.add('Product A', [500, 750, 1000, 1250, 1500])
bar_chart.add('Product B', [750, 1000, 1250, 1500, 1750])

# Render the chart to an SVG and PNG file

> bar_chart.render_to_file('bar_chart.svg')
> drawing = svg2rlg("bar_chart.svg")
> renderPM.drawToFile(drawing, "bar_chart.png")

Reviewer: ```suggestion
svg_img = bar_chart.render()
# Code inspired from: https://github.com/deeplook/svglib/blob/v1.5.1/svglib/svglib.py#L1446
xml_parser = etree.XMLParser(remove_comments=True, recover=True, resolve_entities=resolve_entities)
svg_root = etree.parse(path, parser=xml_parser).getroot()  # from lxml import etree
# Code inspired by: https://github.com/deeplook/svglib/blob/v1.5.1/svglib/svglib.py#L1415
drawing = SvgRenderer('').render(svg_root)  # from svglib.svglib import SvgRenderer
img_bytes = renderPM.drawToString(drawing)
```
Couldn't this be done fully in-memory?

Author: This is clever. Did not think or explore like this. 
I will run it and will let you notify about the verdict.

Author: @Lucas-C , I tried your suggestions. Here are what I found -

- `PyGal` generates SVG strings (XML strings) using the `render()` method.
- The `parse()` method of `lxml` element tree parsing technique can not parse the SVG string as the implementation of `parse()` accepts a `file path` or a `pathlike` object as a parameter instead of an SVG string.
- Instead, I used the `fromstring()` method provided by `lxml` to achieve in-memory SVG parsing.
- The `fpdf.image()` method use `'utf-8' codec` whereas the `renderPM.drawToString()` method returns an image string that might not be a 'utf-8' encoded string which returns following error.
```
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x87 in position 10: invalid start byte
```
I tried to google the error. Turns out the image encoding need to be done in `utf-8` method. And from `fpdf.image()` method's end we can simply ignore that byte which of course not a bright idea here. Link: [Here](https://stackoverflow.com/questions/12468179/unicodedecodeerror-utf8-codec-cant-decode-byte-0x9c). There maybe other approaches to address this issue. I did not dig deeper into this as of now.
- To get rid of this problem we may use the `io.BytesIO` library to make it a Byte string that is readable by the `image()` method. 

Accommodating all the changes of the abovementioned discussion it boils down to the following snippet-

```python
svg_img = bar_chart.render()
xml_parser = etree.XMLParser()
svg_root = etree.fromstring(svg_img, parser=xml_parser)
drawing = SvgRenderer(svg_img).render(svg_root)
drawing_img_byte = renderPM.drawToString(drawing)
img_bytes = io.BytesIO(drawing_img_byte)
```

We can embed charts using this in-memory technique. Please let me know what do you think?

Reviewer: You did a really good job of exploring & experimenting overall! 🙏

`etree.fromstring()` was a really good idea. If it all works with a "bare" `XMLParser` instance, maybe the `parser` argument can be fully omitted? _E.g._ `svg_root = etree.fromstring(svg_img)`. _cf._ [documentation](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.fromstring)

Apart from that minor suggestion, I think this code would be perfect for our new documentation section 😊

# Set the position and size of the image in the PDF
x = 50
y = 50
w = 100
h = 70

pdf = FPDF()
pdf.add_page()
pdf.image('bar_chart.png', x=x, y=y, w=w, h=h)
pdf.output('bar_chart.pdf')

# os.remove('bar_chart.png')  # It removes generated PNG file. Uncomment if needed
# os.remove('bar_chart.svg')  # It removes generate SVG file. Uncomment if needed
```

> The above example takes around 40 seconds to render the image. User who are using `reportlab` and `svglib` to work with `svg` images and are intended to render `svg` images `reportlab` and `svglib` can use this library. However, it does not garranty for a lesser time to accomplish the process.

Reviewer: Those "40 seconds" depends on the user computer hardware & system: is it really meaningful to share it?

Author: You are right. I will remove this line.

Also the above process is not an `in-memory` process as we are required to create both `png` and `svg` to embed into the `fpdf`. 

The above code shows following output -
![](pygal_chart.png)

**Why there is a performance issue between `cairosvg` and `svglib`?**

> *Regarding performance, cairosvg is generally faster than svglib when it comes to rendering SVG files to other formats. This is because cairosvg is built on top of a fast C-based rendering engine, while svglib is written entirely in Python. Additionally, cairosvg offers various options for optimizing the rendering performance, such as disabling certain features, like fonts or filters.*

Reviewer: 👍 Nice explanation!

Author: Thank you .

![](plotly_svg.png)

## Using Pygal ##
[Pygal](https://www.pygal.org/en/stable/) is a graph plotting library using Python. You can install Pygal using `pip install pygal` command.  

> `fpdf2` is able to embed the graph and charts that are generated using `Pygal` library. The following ways explain how to embed `Pygal` charts into `fpdf2` library. However, we can not embed graphs as SVG directly. Since, `Pygal` introduces <style> & <script> elements to the `SVG` images it produces ([Ref](https://github.com/Kozea/pygal/blob/3.0.0/pygal/svg.py#L449)) which currently not supported by `fpdf`. You can see the unsupported SVG features of `fpdf`[here](https://pyfpdf.github.io/fpdf2/SVG.html#currently-unsupported-notable-svg-features).

Reviewer: ```suggestion
`fpdf2` is able to embed the graph and charts that are generated using `Pygal` library. The following ways explain how to embed `Pygal` charts into `fpdf2` library. However, we can not embed graphs as SVG directly. Since, `Pygal` introduces `<style>` & `<script>` elements to the `SVG` images it produces ([Ref](https://github.com/Kozea/pygal/blob/3.0.0/pygal/svg.py#L449)) which is currently not supported by `fpdf2`. The full list of unsupported SVG features of `fpdf2` is [there](https://pyfpdf.github.io/fpdf2/SVG.html#currently-unsupported-notable-svg-features).
```
I think `<style>` & `<script>` should be escaped in backquotes, otherwise the Markdown interpreter may treat them as actual valid HTMl tags! 😊
